runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -9374,6 +9374,13 @@ J9::Z::TreeEvaluator::VMmonentEvaluator(TR::Node * node, TR::CodeGenerator * cg)
    //    CS      monitorReg,GPR13,#lwOffset(objectReg)
    //    BRC     BLRC(0x4), callLabel (OOL path)
 
+   static char* enableNIAIMonEntExit = feGetEnv("TR_enableNIAIMonEntExit");
+   TR::Register* niaiTempReg = cg->allocateRegister();
```

Comment:
Allocation of `niaiTempReg` should be conditional if we are generating that instruction or not. May be we can find a scratch register used in the evaluate for this purpose.

Reply:
Also, once you have made the changes, can you post the instruction selection log got the monent and monexit log here ?

Reply:
Here's what the instruction selection looks like for `monent`:
```
============================================================
; Live regs: GPR=1 FPR=0 VRF=0 {&GPR_0032}
------------------------------
 n18n     (  0)  treetop                                                                              [     0x3ff6e402e90] bci=[-1,3,30] rc=0 vc=151 vn=- li=2 udi=- nc=1
 n17n     (  1)    monent  jitMonitorEntry[#115  helper Method] [flags 0x400 0x0 ]                    [     0x3ff6e402e40] bci=[-1,3,30] rc=1 vc=151 vn=- li=2 udi=- nc=1
 n15n     (  2)      ==>aload (in &GPR_0032) (X!=0 )
------------------------------
monent: Internal Control Flow in OOL : true
ReturnType = NoType
------------------------------
 n18n     (  0)  treetop                                                                              [     0x3ff6e402e90] bci=[-1,3,30] rc=0 vc=151 vn=- li=2 udi=- nc=1
 n17n     (  0)    monent  jitMonitorEntry[#115  helper Method] [flags 0x400 0x0 ]                    [     0x3ff6e402e40] bci=[-1,3,30] rc=0 vc=151 vn=- li=2 udi=- nc=1
 n15n     (  1)      ==>aload (in &GPR_0032) (X!=0 )
------------------------------

 [     0x3ff6e53e180]                          XR      GPR_0035,GPR_0035
 [     0x3ff6e53e470]                          LTG     GPR_0036,#415 4(&GPR_0032)
 [     0x3ff6e53e540]                          NIAI    8,0
 [     0x3ff6e53e6d0]                          CS      GPR_0035,GPR13,#416 4(&GPR_0032)
 [     0x3ff6e53e7b0]                          BRC     BM(0x4), Outlined Label L0033
 [     0x3ff6e561dd0]                          assocreg
 [     0x3ff6e561810]                          Label L0032:
 POST:
 {AssignAny:&GPR_0032:R} {AssignAny:GPR_0035:R} {AssignAny:GPR_0037:R}

------------------------------

 [     0x3ff6e53e940]                          Outlined Label L0033:     ; Denotes start of OOL monent sequence
 [     0x3ff6e53ecd0]                          L       GPR_0035,#417 4(&GPR_0032)
 [     0x3ff6e53eda0]                          LHI     GPR_0037,0xff0f
 [     0x3ff6e53ee80]                          AHI     GPR_0035,0x10
 [     0x3ff6e53ef60]                          NR      GPR_0037,GPR_0035
 [     0x3ff6e53f040]                          CRJ     GPR_0037,GPR13,Label L0036,BNH(mask=0x6),
 [     0x3ff6e53f130]                          ST      GPR_0035,#418 4(&GPR_0032)
 [     0x3ff6e53f200]                          BRC     NOP(0xf), Label L0035
 [     0x3ff6e53f2e0]                          Label L0036:
 [     0x3ff6e55fac0]                          LGR     GPR_0076,GPR13
 [     0x3ff6e55fc50]                          LGR     &GPR_0077,&GPR_0032       ; LR=Clobber_eval
 [     0x3ff6e560b30]                          assocreg
  PRE:
 {GPR2:GPR_0076:R} {GPR3:&GPR_0077:R}
 [     0x3ff6e560540]                          BRASL   GPR_0043,0x0000000000000000
 POST:
 [     0x3ff6e561640]                          assocreg
 [     0x3ff6e561080]                          Label L0035:
 POST:
 {AssignAny:&GPR_0032:R} {AssignAny:GPR_0035:R} {AssignAny:GPR_0037:R} {GPR0:D_GPR_0038:D}* {GPR1:D_GPR_0039:D}* {GPR2:D_GPR_0040:D}* {GPR3:D_GPR_0041:D}* {GPR4:D_GPR_0042:D}*
 {GPR14:GPR_0043:D} {FPR0:D_FPR_0044:D}* {FPR1:D_FPR_0045:D}* {FPR2:D_FPR_0046:D}* {FPR3:D_FPR_0047:D}* {FPR4:D_FPR_0048:D}* {FPR5:D_FPR_0049:D}* {FPR6:D_FPR_0050:D}*
 {FPR7:D_FPR_0051:D}* {FPR8:D_FPR_0052:D}* {FPR9:D_FPR_0053:D}* {FPR10:D_FPR_0054:D}* {FPR11:D_FPR_0055:D}* {FPR12:D_FPR_0056:D}* {FPR13:D_FPR_0057:D}* {FPR14:D_FPR_0058:D}*
 {FPR15:D_FPR_0059:D}* {VRF16:D_VRF_0060:D}* {VRF17:D_VRF_0061:D}* {VRF18:D_VRF_0062:D}* {VRF19:D_VRF_0063:D}* {VRF20:D_VRF_0064:D}* {VRF21:D_VRF_0065:D}* {VRF22:D_VRF_0066:D}*
 {VRF23:D_VRF_0067:D}* {VRF24:D_VRF_0068:D}* {VRF25:D_VRF_0069:D}* {VRF26:D_VRF_0070:D}* {VRF27:D_VRF_0071:D}* {VRF28:D_VRF_0072:D}* {VRF29:D_VRF_0073:D}* {VRF30:D_VRF_0074:D}*
 {VRF31:D_VRF_0075:D}*
 [     0x3ff6e561710]                          BRC     NOP(0xf), Label L0032     ; Denotes end of OOL monent: return to mainline

```

Reply:
And here it is for `monexit`:
```
============================================================
; Live regs: GPR=0 FPR=0 VRF=0 {}
------------------------------
 n53n     (  0)  NULLCHK on n50n [#32]                                                                [     0x3ff6e47e050] bci=[-1,41,39] rc=0 vc=151 vn=- li=5 udi=- nc=1
 n52n     (  1)    monexit  jitMonitorExit[#117  helper Method] [flags 0x400 0x0 ]                    [     0x3ff6e47e000] bci=[-1,41,39] rc=1 vc=151 vn=- li=5 udi=- nc=1
 n50n     (  1)      aload  <auto slot 3>[#401  Auto] [flags 0x7 0x0 ]                                [     0x3ff6e403890] bci=[-1,40,39] rc=1 vc=151 vn=- li=5 udi=- nc=0
------------------------------
aload reg contains ref: 1
Instruction 000003FF6E58B2E0 throws an implicit NPE, node: 000003FF6E47E000 NPE node: 000003FF6E403890
monexit: Internal Control Flow in OOL : true
ReturnType = NoType
monexit: Internal Control Flow in OOL : false
ReturnType = NoType
------------------------------
 n53n     (  0)  NULLCHK on n50n [#32]                                                                [     0x3ff6e47e050] bci=[-1,41,39] rc=0 vc=151 vn=- li=5 udi=- nc=1
 n52n     (  0)    monexit  jitMonitorExit[#117  helper Method] [flags 0x400 0x0 ]                    [     0x3ff6e47e000] bci=[-1,41,39] rc=0 vc=151 vn=- li=5 udi=- nc=1
 n50n     (  0)      aload  <auto slot 3>[#401  Auto] [flags 0x7 0x0 ] (in &GPR_0112) (X!=0 )         [     0x3ff6e403890] bci=[-1,40,39] rc=0 vc=151 vn=- li=5 udi=- nc=0 flg=0x4
------------------------------

 [     0x3ff6e58ac90]                          LG      &GPR_0112, Auto[<auto slot 3>] ?+0(GPR5)
 [     0x3ff6e58b2e0]                          LLZRGF  GPR_0115,#421 0(&GPR_0112) ; Throws Implicit Null Pointer Exception
 [     0x3ff6e58b4e0]                          LTG     GPR_0116,#422 216(GPR_0115)
 [     0x3ff6e58b5b0]                          BRC     BNL(0xc), Outlined Label L0081    ; Branch to OOL monexit monitorLookupCache
 [     0x3ff6e58b750]                          LA      GPR_0116,#423 0(GPR_0116,&GPR_0112)
 [     0x3ff6e58f810]                          LTG     GPR_0158,#428 0(GPR_0116)
 [     0x3ff6e58f8e0]                          NIAI    7,0
 [     0x3ff6e58fa70]                          C       GPR13,#429 0(GPR_0116)
 [     0x3ff6e58fb40]                          BRC     BNH(0x6), Outlined Label L0080
 [     0x3ff6e58fce0]                          MVHI    #430 0(GPR_0116),0x0000
 [     0x3ff6e597d60]                          assocreg
 [     0x3ff6e5977a0]                          Label L0082:
 POST:
 {AssignAny:&GPR_0112:R} {AssignAny:GPR_0114:R} {AssignAny:GPR_0115:R}

------------------------------

 [     0x3ff6e58b9c0]                          Outlined Label L0081:     ; Denotes start of OOL monexit monitorLookupCache
 [     0x3ff6e58bae0]                          RISBGN  GPR_0117,&GPR_0112,57,189,255
 [     0x3ff6e58bc90]                          LLGF    GPR_0116,#424 2128(GPR_0117,GPR13)
 [     0x3ff6e58bd60]                          CGIJ    GPR_0116,Label L0085,0,BH(mask=0x8),
 [     0x3ff6e58bf10]                          CG      &GPR_0112,#425 0(GPR_0116)
 [     0x3ff6e58bfe0]                          BRC     BNH(0x6), Label L0085
 [     0x3ff6e58c180]                          C       GPR13,#426 24(GPR_0116)
 [     0x3ff6e58c250]                          BRC     BNH(0x6), Label L0085
 [     0x3ff6e58c3f0]                          MVHI    #427 24(GPR_0116),0x0000
 [     0x3ff6e58c4d0]                          BRC     NOP(0xf), Label L0086
 [     0x3ff6e58c5b0]                          Label L0085:
 [     0x3ff6e58d8b0]                          LGR     GPR_0156,GPR13
 [     0x3ff6e58da40]                          LGR     &GPR_0157,&GPR_0112       ; LR=Clobber_eval
 [     0x3ff6e58e9c0]                          assocreg
  PRE:
 {GPR2:GPR_0156:R} {GPR3:&GPR_0157:R}
 [     0x3ff6e58e3d0]                          BRASL   GPR_0123,0x0000000000000000
 POST:
 [     0x3ff6e58f4f0]                          assocreg
 [     0x3ff6e58ef30]                          Label L0086:
 POST:
 {AssignAny:&GPR_0112:R} {AssignAny:GPR_0114:R} {AssignAny:GPR_0116:R} {AssignAny:GPR_0117:R} {GPR0:D_GPR_0118:D}* {GPR1:D_GPR_0119:D}* {GPR2:D_GPR_0120:D}* {GPR3:D_GPR_0121:D}*
 {GPR4:D_GPR_0122:D}* {GPR14:GPR_0123:D} {FPR0:D_FPR_0124:D}* {FPR1:D_FPR_0125:D}* {FPR2:D_FPR_0126:D}* {FPR3:D_FPR_0127:D}* {FPR4:D_FPR_0128:D}* {FPR5:D_FPR_0129:D}*
 {FPR6:D_FPR_0130:D}* {FPR7:D_FPR_0131:D}* {FPR8:D_FPR_0132:D}* {FPR9:D_FPR_0133:D}* {FPR10:D_FPR_0134:D}* {FPR11:D_FPR_0135:D}* {FPR12:D_FPR_0136:D}* {FPR13:D_FPR_0137:D}*
 {FPR14:D_FPR_0138:D}* {FPR15:D_FPR_0139:D}* {VRF16:D_VRF_0140:D}* {VRF17:D_VRF_0141:D}* {VRF18:D_VRF_0142:D}* {VRF19:D_VRF_0143:D}* {VRF20:D_VRF_0144:D}* {VRF21:D_VRF_0145:D}*
 {VRF22:D_VRF_0146:D}* {VRF23:D_VRF_0147:D}* {VRF24:D_VRF_0148:D}* {VRF25:D_VRF_0149:D}* {VRF26:D_VRF_0150:D}* {VRF27:D_VRF_0151:D}* {VRF28:D_VRF_0152:D}* {VRF29:D_VRF_0153:D}*
 {VRF30:D_VRF_0154:D}* {VRF31:D_VRF_0155:D}*
 [     0x3ff6e58f5c0]                          BRC     NOP(0xf), Label L0082     ; Denotes end of OOL monexit monitorCacheLookup: return to mainline
 [     0x3ff6e58fe10]                          Outlined Label L0080:
 [     0x3ff6e595910]                          LGR     GPR_0197,GPR13
 [     0x3ff6e595aa0]                          LGR     &GPR_0198,&GPR_0112       ; LR=Clobber_eval
 [     0x3ff6e596570]                          assocreg
 [     0x3ff6e596f40]                          assocreg
  PRE:
 {GPR2:GPR_0197:R} {GPR3:&GPR_0198:R}
 [     0x3ff6e595f80]                          BRASL   GPR_0164,0x0000000000000000
 POST:
 {GPR0:D_GPR_0159:D}* {GPR1:D_GPR_0160:D}* {GPR2:D_GPR_0161:D}* {GPR3:D_GPR_0162:D}* {GPR4:D_GPR_0163:D}* {GPR14:GPR_0164:D} {FPR0:D_FPR_0165:D}* {FPR1:D_FPR_0166:D}*
 {FPR2:D_FPR_0167:D}* {FPR3:D_FPR_0168:D}* {FPR4:D_FPR_0169:D}* {FPR5:D_FPR_0170:D}* {FPR6:D_FPR_0171:D}* {FPR7:D_FPR_0172:D}* {FPR8:D_FPR_0173:D}* {FPR9:D_FPR_0174:D}*
 {FPR10:D_FPR_0175:D}* {FPR11:D_FPR_0176:D}* {FPR12:D_FPR_0177:D}* {FPR13:D_FPR_0178:D}* {FPR14:D_FPR_0179:D}* {FPR15:D_FPR_0180:D}* {VRF16:D_VRF_0181:D}* {VRF17:D_VRF_0182:D}*
 {VRF18:D_VRF_0183:D}* {VRF19:D_VRF_0184:D}* {VRF20:D_VRF_0185:D}* {VRF21:D_VRF_0186:D}* {VRF22:D_VRF_0187:D}* {VRF23:D_VRF_0188:D}* {VRF24:D_VRF_0189:D}* {VRF25:D_VRF_0190:D}*
 {VRF26:D_VRF_0191:D}* {VRF27:D_VRF_0192:D}* {VRF28:D_VRF_0193:D}* {VRF29:D_VRF_0194:D}* {VRF30:D_VRF_0195:D}* {VRF31:D_VRF_0196:D}*
 [     0x3ff6e5975d0]                          assocreg
 [     0x3ff6e597010]                          Label L0084:
 POST:
 {AssignAny:&GPR_0112:R} {AssignAny:GPR_0114:R} {AssignAny:GPR_0115:R}
 [     0x3ff6e5976a0]                          BRC     NOP(0xf), Label L0082     ; Denotes end of OOL monexit: return to mainline

Reply:
[niai_log.txt](https://github.com/eclipse-openj9/openj9/files/14622092/niai_log.txt)
Attaching the log here as well for reference.